clmock/clmock.go

```diff
@@ -0,0 +1,222 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package clmock
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type CLMock struct {
+	ctx               context.Context
+	cancel            context.CancelFunc
+	stack             *node.Node
+	eth               *eth.Ethereum
+	blockPeriod       time.Duration
+	withdrawals       []*types.Withdrawal
+	feeRecipient      common.Address
+	feeRecipientMutex sync.Mutex
+	nextWithdrawalIdx uint64
+}
+
+func NewCLMock(stack *node.Node, eth *eth.Ethereum) *CLMock {
+	chainConfig := eth.APIBackend.ChainConfig()
+
+	if chainConfig.Dev == nil {
+		log.Crit("incompatible pre-existing chain configuration")
+	}
+
+	return &CLMock{
+		stack:        stack,
+		eth:          eth,
+		blockPeriod:  time.Duration(chainConfig.Dev.Period) * time.Second,
+		withdrawals:  []*types.Withdrawal{},
+		feeRecipient: common.Address{},
+	}
+}
+
+// Start invokes the clmock life-cycle function in a goroutine
+func (c *CLMock) Start() error {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.clmockLoop()
+	return nil
+}
+
+// Stop halts the clmock service
+func (c *CLMock) Stop() error {
+	c.cancel()
+	return nil
+}
+
+func (c *CLMock) addWithdrawal(w types.Withdrawal) error {
+	if w.Index < c.nextWithdrawalIdx {
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", w.Index, c.nextWithdrawalIdx-1)
+	}
+	c.nextWithdrawalIdx = w.Index + 1
+	c.withdrawals = append(c.withdrawals, &w)
+	return nil
+}
+
+// remove up to 10 withdrawals from the withdrawal queue
+func (c *CLMock) popWithdrawals() []*types.Withdrawal {
+	var popCount int
+	if len(c.withdrawals) >= 10 {
+		popCount = 10
+	} else {
+		popCount = len(c.withdrawals)
+	}
+
+	popped := make([]*types.Withdrawal, popCount)
+	copy(popped[:], c.withdrawals[0:popCount])
+	c.withdrawals = append([]*types.Withdrawal{}, c.withdrawals[popCount:]...)
+	return popped
+}
+
+// clmockLoop manages the lifecycle of clmock.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *CLMock) clmockLoop() {
+	ticker := time.NewTicker(time.Millisecond * 100)
+	lastBlockTime := time.Now()
+
+	var curForkchoiceState engine.ForkchoiceStateV1
+
+	engineAPI := catalyst.NewConsensusAPI(c.eth)
+
+	header := c.eth.BlockChain().CurrentHeader()
+
+	curForkchoiceState = engine.ForkchoiceStateV1{
+		HeadBlockHash:      header.Hash(),
+		SafeBlockHash:      header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
```

Comment:
should probably just put all these in a `var` block - also no need to forward declare `curForkchoiceState`